flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java

```diff
@@ -193,6 +202,11 @@ private CheckpointBarrierBehaviourController chooseController(CheckpointBarrier
 
 	private boolean canTimeout(CheckpointBarrier barrier) {
 		return barrier.getCheckpointOptions().isTimeoutable() &&
-			barrier.getCheckpointOptions().getAlignmentTimeout() < (System.currentTimeMillis() - barrier.getTimestamp());
+			barrier.getId() <= lastSeenBarrier &&
+			barrier.getCheckpointOptions().getAlignmentTimeout() * 1_000_000 < (System.nanoTime() - firstBarrierArrivalTime);
+	}
+
+	private long getArrivalTime(CheckpointBarrier announcedBarrier) {
+		return announcedBarrier.getCheckpointOptions().isTimeoutable() ? System.nanoTime() : Long.MAX_VALUE;
```

Comment:
I'm not entirely convinced if this a better approach. 

For me, using global checkpoint time (in other words recently added checkpoinStartDelay metric) seemed easier to understand for the user. If aligned checkpoint barrier is taking too much time to reach given tasks - we are timeouting it to unaligned checkpoint that overtakes in-flight data. This seemed easier to comprehend and easier to explain  compared to time outing alignment on some subtask?

Secondly your proposed change will not work with single input tasks without active timeouts?

Reply:
>  easier to understand

I agree that it might be true for some users, but not for all. During the [previous discussion](https://github.com/apache/flink/pull/13827#discussion_r527794600), and also the one before, the consensus was that it's **not** easier to understand. However, we can discuss it again. 
(also there are some more technical advantages of "local" timeouts)

> Secondly your proposed change will not work with single input tasks without active timeouts?

Why, could you explain?

Reply:
> Why, could you explain?

On a second thought, maybe it will partially, but not fully as well.  guess we still have the code, that we can timeout to UC on the last processed barrier? So in case of single channel:
1. announcement is processed (first announcement will never timeout in this version) it won't timeout
2. barrier will be processed, and only it can timeout

So a timeout on the first network exchange will work worse. That's a bit problematic, especially for simple jobs, with for example just a single exchange. Previous version would cut the checkpointing time by half, this version will do worse than that, in a way that's hard to quantify for me. 

There is some extreme corner case when imagine there is a heavy back pressure, but all CB are processed at the same time. That means announcements in this version wouldn't cause timeout (it would in my older proposal), and this version will need to wait for some CB to be processed (which can take long time).

Active timeout would alleviate this problem though.

Reply:
With a single input, this is not an alignment issue but a (related) back-pressure issue. And I think this is the right way to solve it: 
> Active timeout would alleviate this problem though.